### ISSUE TITLE
Optional chaining for arrays inferring wrong type

### ISSUE THREAD

**Issue by dbersan** [microsoft/TypeScript#58033]
### 🔎 Search Terms

"Optional Chaining", "arrays",

### 🕗 Version & Regression Information

- Happens on version 5.4.3

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMpgPZRQbwFDKHIAOUGAJgK4JgDOAXMjssOY7WFKAOYDcJXJIxCUAtgCNoyAL4BtALq880vHgQYQHZDAwZkAXmQAKWhlEQOWCI3RWAlAYB8TAkXWawLcgHkYAMWAoDgAFMioaA2RTc0tsADpSCmo6AH442QAGeTTWfgB6POQAHgBaEpYQGGhaZDAAT2IUTGRJZAADDi4QbjaKjgg4cmQMGHbOnmQAH2RKEHIIGFAIcjbXQmAAJQgkYAA3CHQu7lojVl8AoLBQpJo7JWklNQ0tTe2IPYPOHhrDEzMLTjsL7dBz6Zz4IjIdymAA2EDiMIw3D+MU4d2Uj2hnlEdVs2HoeJQhhwiXCdHoChUeB0GCMOMJdlUhCAA

### 💻 Code

```ts
interface Store {
    products: { id: string; price: number }[];
}

const foo = (somestore: Store) => {
    const idOfFirstProduct = somestore.products?.[0]?.id;  //<- idOfFirstProduct SHOULD be string | undefined
    iReceiveStrings(idOfFirstProduct);
};

const iReceiveStrings = (somestr: string) => {
    console.log(somestr);
};

const myStore:Store = {products:[]}

foo(myStore) // <-- error not caught by typescript!
```

### 🙁 Actual behavior

Typescript does not infer the type should be possibly `undefined` when using optional chaining on arrays

### 🙂 Expected behavior

Typescript should infer the type should be possibly `undefined` when using optional chaining on arrays

**Comment by MartinJohns** [comment]
This is working as intended. According to your types `products` can't ever be nullish, so you won't get `undefined` here.

**Comment by snarbies** [comment]
Enabling `noUncheckedIndexedAccess` produces the expected error for undefined for your array access.

    Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
    Type 'undefined' is not assignable to type 'string'.(2345)
